keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,149 @@
+
+    def preprocess_features(self, x):
+        """An overridable function which preprocesses features."""
+        return x
+
+    def preprocess_labels(self, y):
+        """An overridable function which preprocesses labels."""
+        return y
+
+    def preprocess(self, *args):
+        """An overridable function which preprocesses samples."""
+        x, y, sample_weight = keras.utils.unpack_x_y_sample_weight(args)
+        x, y = self.preprocess_features(x), self.preprocess_labels(y)
+        return keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+
+    # ========================================================================
+    # Below are overrides to keras.Model methods to apply the functions above.
+    # ========================================================================
+    def fit(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        validation_data=None,
+        validation_split=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess)
+
+        if validation_split:
+            raise ValueError(
+                "`validation_split` is not supported on this model. "
+                f"Received: `validation_split={validation_split}`."
+            )
+        if validation_data is not None:
+            val_tuple = keras.utils.unpack_x_y_sample_weight(validation_data)
+            validation_data = convert_inputs_to_dataset(*val_tuple, batch_size)
+            if self.include_preprocessing:
+                validation_data = apply_dataset_fn(
+                    validation_data, self.preprocess
+                )
+
+        return super().fit(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            validation_data=validation_data,
+            **kwargs,
+        )
+
+    def evaluate(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess)
+
+        return super().evaluate(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            **kwargs,
+        )
+
+    def predict(
+        self,
+        x=None,
+        batch_size=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, None, None, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess_features)
+
+        return super().predict(
+            x=x,
+            batch_size=None,
+            **kwargs,
+        )
+
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
+        if include_preprocessing is None:
```

Comment:
Let's try just making this dtype aware to avoid preprocessing what has already been preprocessed.